---
title: "Using crossr for Quantitative Comparison of Transcriptomes Between Species"
author: "Otho Mantegazza"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Crossr Workflow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction
In theory all the differences between two species are encoded in their genomes. Nevertheless, at the present day, it is hard to interpret the information contained in a genome without external aid. Measurement of steady state transcript level (today mostly done through RNAseq) across different conditions can help making sense of the information contained in a genome. While the methods for quantitative comparison of transcript level among samples originating from the same species are widely developed and available, we cannot say the same when the samples come from different species. There are multiple issues that make this task not straightforward. Here we present a practical workflow that we use for comparing gene expression among different species.
The workflow builds up on the outputs of Salmon for quantifying RNAseq reads and on Orthofinder for inferring homology groups, therefore requires that a reference transcriptome is available for both the species. While developing this workflow we focused on simplicity and usability, therefore this has to be intended as a simple method for observing potentially interesting differences in gene expression among species rather then as the best method to draw conclusion on such differences.
