---
title: "Using crossr for Quantitative Comparison of Transcriptomes Between Species"
author: "Otho Mantegazza"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Crossr Workflow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction
In theory all the differences between two species are encoded in their genomes. Nevertheless, at the present day, it is hard to interpret the information contained in a genome without external aid. Measurement of steady state transcript level (today mostly done through RNAseq) across different conditions can help making sense of the information contained in a genome.

While the methods for quantitative comparison of transcript level among samples originating from the same species are widely developed and available, we cannot say the same when the samples come from different species. There are multiple issues that make this task not straightforward. Here we present a practical workflow that I use for comparing gene expression among different species.

The workflow builds up on the outputs of Salmon, for quantifying RNAseq reads, and on Orthofinder, for inferring homology groups; therefore a reference transcriptome is required to run this workflow

During development, I focused on simplicity and usability. Hence, This has to be intended as a simple method for observing potentially interesting differences in gene expression among species rather then as the best method to draw conclusion on such differences.

# Rationale 

# Workflow

Since I produced this package for my practical use, I think that the function contained in it and its workflow are better explained with a practical example. 

Hereby I use the *Arabidopsis lyrata* vs *Arabidopsis thaliana* dataset produced **Pietzenuk et al.** in **2015** and published in PLoS One under the title [Improving the Annotation of Arabidopsis lyrata Using RNA-Seq Data](http://dx.doi.org/10.1371/journal.pone.0137391).

In this work, the authors compare the transcriptomes of rosette leaves from both the species before, during and after an heatshock. The authors made the raw and processed data available on the [Short Reads Archive](https://www.ncbi.nlm.nih.gov/sra/) under the GEO accession [GSE69077](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE69077)

## Estimation of Transcript abundance (Salmon)

I generally use [Salmon](https://combine-lab.github.io/salmon/) for quantifying RNAseq data because it is fast, accurate and easy to use. Any other aligner / pseudoaligner could anyway be adapted to fit the workflow.

Salmon provides **Transcripts Per Millions (TPM)** directly as expression measure. TPM are not dependent on gene length. In my experience this is an helpful feature when dealing with cross species comparison because it is independent of gene length. As we will see later, often homology groups contain genes of very different length.

### Loading the Salmon output

When running Salmon on a set of FASTQ samples, the output is a generally a set of folders (one for any FASTQ file). I suggest to organize collect all the output subfolders from one species in one folder and to extract the TPM with the function `make_TPM_df()` which is provided by the package.

For example, here all the outputs for the *Arabidopsis thaliana* samples are organized in the `thaliana` folder

```{r, echo=FALSE, message=FALSE}
devtools::load_all()
```

```{r}
list.files("../genomes/thaliana")
```

and every subfolder contain a typical Salmon output

```{r}
list.files("../genomes/thaliana/SRR2033948")
```

The function `make_TPM_df` extracts the TPM from all the `.sf` files in every subfolder and organizes them in a `data.frame`.

```{r}
thaliana <- make_TPM_df("../genomes/thaliana")
str(thaliana)
```

## Session info

Here is the output of `devtools::session_info()` on the system on which this document was compiled:

```{r, echo=FALSE}
devtools::session_info()
```



