---
title: "A quick introduction to crossr"
author: "Otho Mantegazza"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{overview of crossr}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo=FALSE, message=FALSE}
devtools::load_all()
```

Crossr is a set of function and a workflow that I use for quantitative comparison of gene expression in different species.

The package builds up on the output of [Salmon](https://combine-lab.github.io/salmon/) for transcript quantification and of [Orthofinder](https://github.com/davidemms/OrthoFinder) for matching homologues.

In this vignette I replicate the crossr workflow on this dataset: [GSE69077](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE69077), which was published by [Rawat et al. in 2015](https://www.ncbi.nlm.nih.gov/pubmed/26382944)

In order to make this Vignette lighter, the workflow is run here on a random subset of 500 orthogroups.

## Load Data
 
The fucntion `make_TPM_df()` loads the Salmon output in R. It takes as input the path to the folder that contains all the Salmon output subfolders for a species

```{r}
list.files("../inst/exdata/thaliana")
```

```{r}
list.files("../inst/exdata/thaliana/SRR2033948")
```

The function `make_TPM_df` extracts the TPM from all the `.sf` files in every subfolder and organizes them in a `data.frame`.

Here I use it to produce an expression matrix for the two species `thaliana` and `lyrata`.

```{r}
thaliana <- make_TPM_df("../inst/exdata/thaliana")
str(thaliana)
dim(thaliana)
```

```{r}
lyrata <- make_TPM_df("../inst/exdata/lyrata")
str(lyrata)
dim(lyrata)
```

if more info on the samples are available, it might be useful to change the `colnames` of the expression sets to a more informative version. In this case we can use the feature table to do so.

```{r}
sample_info <-  read.table("../inst/exdata/SRP058527_metadata.txt",
                           header = TRUE,
                           sep = "\t")  
reduce_info <- function(info) {
    if(info == "heat stress 0 h") "0h"
    else if(info == "heat stress 6 h 37\302\260C") "heat_6h"
    else "recovery"
}
sample_info$etime <- vapply(sample_info$exposure_time_s,
                            reduce_info,
                            character(1))
sample_info$etime <- paste(sample_info$etime,
                           c(1, 2),
                           sep = "_")

switch_name <- function(name) {
    if(name %in% sample_info$Run_s) {
    sample_info[sample_info$Run_s == name, "etime", drop = TRUE]
    } else name
    }
colnames(thaliana) <- vapply(colnames(thaliana), switch_name, character(1))
colnames(thaliana) <- paste("thal",
                            colnames(thaliana),
                            sep = "_")
str(thaliana)

colnames(lyrata) <- vapply(colnames(lyrata), switch_name, character(1))
colnames(lyrata) <- paste("lyr",
                          colnames(lyrata),
                          sep = "_")
str(lyrata)
```


## Load orthogroups

The homology groups are called orthogroups and in [orthofinder](https://github.com/davidemms/OrthoFinder) are defined as:

> the set of genes that are descended from a single gene in the last common ancestor of all the species being considered.

The orthogroups are provided in the `OrthologousGroups.csv` file in the orthofinder output.
Orthofinder provides both an `OrthologousGroups.csv` file and `OrthologousGroups.txt` file. The function `parse_orthogroups()` takes as input **exclusively the csv file** provided as a path.

```{r}
ogroups <- parse_orthogroups("../inst/exdata/ogroups_sample.csv")
```

The dimension of the orthogroups can be explored with the function `explore_ogroups()`

```{r, fig.width= 7, fig.height=4, fig.cap="Figure 1"}
explore_ogroups(ogroups)
```


## Convert protein ID to transcript ID

Since orthogroups are inferred on protein sequences but short reads are mapped on transcript sequences you might have to convert the protein ID in the orthogroups in transcript ID. The function `switch_ids()` performs this task. It requires an external file that maps the protein ID to transcript ID, which must be loaded in R.

Loading the feature table from refseq can be challenging, this loads the *A. thaliana* one
```{r}
thaliana_id_path <- "../inst/exdata/thaliana_features_sample.txt"
thaliana_ids <- read.table(file = thaliana_id_path,
                           sep = "\t", header = FALSE,
                           stringsAsFactors = FALSE,
                           quote = "")

## Not sure why the header is commented with "#"
colnames(thaliana_ids) <-  scan(file = thaliana_id_path,
                                what = "",
                                nlines = 1)[-1]  
```

And this loads the *A. lyrata* one.
```{r}
lyrata_id_path <- "../inst/exdata/lyrata_features_sample.txt"
lyrata_ids <- read.table(file = lyrata_id_path,
                         sep = "\t", header = FALSE,
                         stringsAsFactors = FALSE,
                         quote = "",
                         skip = 1,
                         comment.char = "") # One gene name contains "#"

## Not sure why the header is commented with "#"
colnames(lyrata_ids) <-  scan(file = lyrata_id_path,
                                what = "",
                                nlines = 1)[-1]  
```


Then we can use `switch_ids()` to convert the IDS from *A. thaliana* first, 

```{r}
ogroups <- switch_ids(ogroups = ogroups,
                      ids_table = thaliana_ids,
                      px_id = "product_accession",
                      tx_id = "related_accession",
                      mc.cores = 4)
```

and then the ones from *A. lyrata*.

```{r}
ogroups <- switch_ids(ogroups = ogroups,
                      ids_table = lyrata_ids,
                      px_id = "product_accession",
                      tx_id = "related_accession",
                      mc.cores = 4)
```

## Implement a S4 element

Now that we have loaded the expression data and the orthogroup data, we can start implementing an S4 element to organize and link together all the various expression set, annotoations and statistics that we have loaded and/or that we are going to produce.

If reader want an introduction to S4 classes in R, I suggest this chapter from Hadley Wickham's advanced R book: http://adv-r.had.co.nz/S4.html

The *crossr* package defines the S4 class `ogset` which can be created with thee `make_ogset` constructor function.

```{r}
og_set <-  make_ogset(og = ogroups,
                       spec1_exp = thaliana,
                       spec2_exp = lyrata)
```

The `ogset` class is made to store orthogroup expression, and it has slots for orthogroup expression, list of genes in orthogroups, single species expression data, design and annotation data. But not all these elements are not necessary when initialising the object with `make_ogset()`. The `ogset` class is modeled on Bioconductor Summarized experiment class.

When the `ogset` element is initialized, the function `check_ogset()` is called in order to check if the elements of the object match with each others. In the specific, `check_ogset()` checks that:

- The names of the orthogroups match the rownames of the orthogroup expression set `og_eset` (if any is given),
- The row names of `colData` match the column names of the orthogroup expression set,
- The row names of `rowData`match the row names of the orthogroup expression set.
- The variable in `design` are colnames of `colData`.
- The sample information (colData) of the two species expression dataframes contain the same experimental variables.

After adding new elements to an `ogset` object, one can check that the object is still valid by calling `check_ogset()` which returns TRUE if the object is valid.

```{r}
check_ogset(og_set)
```

### Save sample information (colData) in the ogset element

We specify the sample information in the `spec1_colData` and `spec2_colData` slots of the `ogset` element.

In this case we have already expressed the sample information in the colnames of the two datasets, and we can use that to specify very basic colData, but more complex and detailed sample information can be provided.

```{r}
colnames(og_set@spec1_exp)
colnames(og_set@spec2_exp)

coldata_spec1 <- sapply(colnames(og_set@spec1_exp), function(i) strsplit(i, split = "_")[[1]][2])
coldata_spec1 <- data.frame(treat = coldata_spec1, stringsAsFactors = FALSE)
coldata_spec1

coldata_spec2 <- sapply(colnames(og_set@spec2_exp), function(i) strsplit(i, split = "_")[[1]][2])
coldata_spec2 <- data.frame(treat = coldata_spec2, stringsAsFactors = FALSE)
coldata_spec2

og_set@spec1_colData <- coldata_spec1
og_set@spec2_colData <- coldata_spec2
``` 

then we have to provide information on which column of the `coldata_spec` datasets contains the main information for the experimental condition that is used to compare the two species. The levels of this column must be the same for both datasets; which means that we are testing for the response to the same experimental condition in the two species.

```{r}
og_set@exp_cond <- "treat"
```

and we can check that the ogset element is still valid

```{r}
check_ogset(og_set)
```


## Collapse orthologues

As seen in Figure 1 most of the orthogroup contain only two genes, generally one from lyrata and one thaliana. Many other orthogroups though contain different (any) number of genes from one species and from the other. A "quick and dirty"" solution to restore a one to one relationship among features is to collapse the orthologues by adding up the expression values of genes within the same orthogroup.

The basic **assumption** of this method is that **genes in the same orthogroup have the same function**.

The function `collapse_orthologs()`performs this task on the `ogset` class elements, and returns the same element with the collapsed orthogroup expression set in the designated slot.

```{r}
og_set <- collapse_orthologs(og_set, mc.cores = 4)
```

Not all orthogroups contain genes from both species, the orthogroups that contain genes and hence expression data from both species are stored in the slot `og_exp` for further analysis,

```{r}
str(og_set@og_exp)
```

the ones that does not are stored in the slot `og_nomatch`

```{r}
str(og_set@og_nomatch)
```

After we have modified the ogset element (anytime), we can check again for its consistency with the function `check_ogset()` (optional).

```{r}
check_ogset(og_set)
```


## Select genes that behave differently between species

Now that we moved to ortholog-wise expression, we are dealing with a one to one relationship among species.

In order to stay on the safe side I do not perform direct comparison of expression level between the two species, but use the F-value on the interaction term to screen for orthogroups that display a relative different expression pattern between the two species.

Also, since the data have gone through many transformation it is hard to make assumption on their distribution, therefore I do not extract a p-value, rather I use the F-value on the interaction term as a ranking feature, in order to select the genes that seem to behave differently among the species.

The function that fits the linear model and performs ANOVA is called `add_fit()`

----------

First we have to add the column data (the information on the sample) to the ogset element and to provide it with details on the experimental design.

The column data were previoulsy encoded in the colnames of the two expression sets from the two different species. Those names are preserved in the colnames of the ortholog expression set.

```{r}
colnames(og_set@og_exp)
```

So I organize those info in a dataframe:

```{r}
coldata <- data.frame(spc = sapply(colnames(og_set@og_exp),
                                   function(i) strsplit(i, split = "_")[[1]][1]),
                      treat = sapply(colnames(og_set@og_exp),
                                     function(i) strsplit(i, split = "_")[[1]][2]),
                      stringsAsFactors = FALSE)

coldata
```

And provide those information to the ogset element

```{r}
og_set@colData <- coldata
check_ogset(og_set)
```

The experimental design must also be provided, indicating which variable to compare:

```{r}
og_set@design <- ~ 0 + spc + treat + spc:treat
check_ogset(og_set)
```

----------

Now we can calculate the F-values. 

I would suggest to do so on log scaled value, since this should be more appropriate for studying relative changes. This could be done by passing the argument `log_scale = TRUE` to the `add_fit()` function (actually 1 is added to the data before natural log trasformation, so the transformation performed is `log(data + 1)`.

```{r}
og_set <- add_fit(ogset = og_set, log_scale = TRUE)
```

The info on the `log_scale` parameter are stored in the `metadata` slot

## Explore results

We have an F-value associated to any of the three terms in the design formula that we have supplied

```{r}
head(og_set@stats)
```

The F statistic for adding the interaction term to the linear model (in this case `spc:treat`) can be used as a feature that ranks orthogroups in base of how likely they are to behave differently between different species. The ID of the orthogroups with the highest F-value can be extracted with the function `get_top_tags()`.

```{r}
tt <- get_top_tags(og_set, "spc:treat", 10)
tt
```

`get_top_tags()` returns a named numeric vector containing both the id of the top orthogroups and their associated statistics.

### Explore orthogroup expression

The function `plot_all_stages` can be used to plot a stripchart of the expression level of one orthogroups in all the samples.
It requires as input:

- the name of the orthogroup,
- the the ogset element,
- the species encoding in the design formula (in this case `spc`)
- the condition encoding in the design formula (in this case `treat`)

```{r, fig.height=4, fig.width=5}
plot_all_stages(orthogroup = "OG0016450",
                ogset = og_set,
                species = "spc",
                condition = "treat",
                ylab = "Log Scaled TPM")
```

This can be combined with faceting in order to plot the expression of multiple orthogroups. For example here I plot the top 12 differentially expressed orthogroups in a 4X3 faceting

```{r, fig.height=9, fig.width=7}
tt <- names(get_top_tags(og_set, "spc:treat", 12))
par(mfrow = c(4, 3))
sapply(tt,
       plot_all_stages,
       ogset = og_set,
       species = "spc",
       condition = "treat",
       ylab = "Log Scaled TPM")
```

### Plot all the genes in one orthogroup

The workflw of `crossr` relies on the assumption that:

> all the genes in one orthogroup have the same function. 

Therefore their exoression values can be added without this resulting in a misenterpretation of the activity of certain groups.

This is a useful working assumption but it is not always verified, therefore I suggest, also, before drawing any conclusion, to explore also the expression pattern of the single genes contained in 1 orthogroup.

This can be achieved with the function `plot_og_genes`.

```{r}
cdata_lyrata <- sapply(colnames(lyrata), function(i) strsplit(i, split = "_")[[1]][2])
cdata_thaliana <- sapply(colnames(thaliana), function(i) strsplit(i, split = "_")[[1]][2])

plot_og_genes(ogroup = rownames(og_fit)[1],
              og_list = ogroups,
              eset_spec1 = thaliana,
              eset_spec2 = lyrata,
              coldata_spec1 = cdata_thaliana,
              coldata_spec2 = cdata_lyrata,
              ylab = "TPM")
```

As we can see the orthogroup `r rownames(og_fit)[1]` contains only 2 genes, but we can expect orthogroups of different size and different ratios between genes from ine species and from the other.

### Example, orthogroup OG0002779

```{r}
example_group <- "OG0002779"
# also OG0000214 makes a good example
```


Orthogroup `r example_group` contains contains 4 genes, 2 from each species.

The F-value associated to this group is `r og_fit[example_group, "spc:treat"]`, and it ranks `r which(rownames(og_fit) == example_group)` out of 500.

Looks like it is behaving differently in response to heat in the 2 species;

```{r, fig.height=4, fig.width=5}
plot_all_stages(orthogroup = example_group,
                dset = og_eset_log,
                species = coldata$spc,
                condition = coldata$treat,
                ylab = "Log Scaled TPM")
```

and it contains 4 genes of which one is behaving peculiarly differently than the others.

```{r, fig.height=6, fig.width=6}
par(mfrow = c(2, 2))
plot_og_genes(ogroup = example_group,
              og_list = ogroups,
              eset_spec1 = thaliana,
              eset_spec2 = lyrata,
              coldata_spec1 = cdata_thaliana,
              coldata_spec2 = cdata_lyrata,
              ylab = "TPM")
```

## Clusters

Ranking orthogroups by F-value of the interaction term and arbitrary cutoffs can be useful to detect genes that behave differently in the two species, but in this way we select genes that behave differently in any way.

Clusters and heatmap can be useful in order to group together orthogroups that behave differently in a similar way and to make sense of it all.



## Session info

Here is the output of `devtools::session_info()` on the system on which this document was compiled:

```{r, echo=FALSE}
devtools::session_info()
```

